generators/beamAndVertexGenerator.cc

```diff
@@ -118,80 +201,201 @@ void beamAndVertexGenerator::randomizeBeamfileStartingPosition() {
-			_beamTree->GetEntry(_currentBeamfileEntry);
-		} else {
-			if(_currentBeamfileEntry >= nEntries) {
-				printInfo << "reached end of beamfile, looping back to first event." << endl;
-				_currentBeamfileEntry = 0;
-			}
-			_beamTree->GetEntry(_currentBeamfileEntry++);
-		}
-		double dx = _beamMomentumX.first / _beamMomentumZ.first;
-		double dy = _beamMomentumY.first / _beamMomentumZ.first;
-		double projectedVertexX = _vertexX.first + dx * z;
-		double projectedVertexY = _vertexY.first + dy * z;
-		if(_sigmasPresent and _sigmaScalingFactor != 0.) {
-			_vertex.SetXYZ(randomGen->Gaus(projectedVertexX, _sigmaScalingFactor * _vertexX.second),
-			               randomGen->Gaus(projectedVertexY, _sigmaScalingFactor * _vertexY.second),
-			               z);
-			_beam.SetXYZM(randomGen->Gaus(_beamMomentumX.first, _sigmaScalingFactor * _beamMomentumX.second),
-			              randomGen->Gaus(_beamMomentumY.first, _sigmaScalingFactor * _beamMomentumY.second),
-			              randomGen->Gaus(_beamMomentumZ.first, _sigmaScalingFactor * _beamMomentumZ.second),
-						  beam.particle.mass());
-		} else {
-			_vertex.SetXYZ(projectedVertexX, projectedVertexY, z);
-			_beam.SetXYZM(_beamMomentumX.first,
-			              _beamMomentumY.first,
-			              _beamMomentumZ.first,
-			              beam.particle.mass());
-		}
+		_vertex.SetXYZ(projectedVertexX, projectedVertexY, z);
+		_beam.SetXYZM(_beamMomentumX,
+						_beamMomentumY,
+						_beamMomentumZ,
+						beam.particle.mass());
 	}
 	return true;
 }
 
 
-double beamAndVertexGenerator::getVertexZ(const Target& target) const {
+bool
+beamAndVertexGenerator::eventFromMomentaCorrectMomentumResolution(const Target& target, const Beam& beam) {
+	TRandom3* randomGen = randomNumberGenerator::instance()->getGenerator();
+	if(not loadNextEventFromBeamfile()) return false;
+
+	const double z = getVertexZ(target); // has to be done AFTER loading the next event
+
+	const double dx = _beamMomentumX / _beamMomentumZ;
+	const double dy = _beamMomentumY / _beamMomentumZ;
+	const double projectedVertexX = _vertexX + dx * (z-_vertexZ);
+	const double projectedVertexY = _vertexY + dy * (z-_vertexZ);
+
+	_beamMomentum = sqrt(_beamMomentumX*_beamMomentumX + _beamMomentumY*_beamMomentumY + _beamMomentumZ*_beamMomentumZ);
+	const double truthMomentumPDFSigma = sqrt(1.0/(1.0/_momentumResolution/_momentumResolution+ 1.0/_momentumPDFSigma/_momentumPDFSigma));
+	const double truthMomentumPDFMean = truthMomentumPDFSigma*truthMomentumPDFSigma*( 1./_momentumResolution/_momentumResolution*_beamMomentum
+																					+ 1./_momentumPDFSigma/_momentumPDFSigma*_momentumPDFMean );
+	const double truthMomentum = randomGen->Gaus(truthMomentumPDFMean, truthMomentumPDFSigma);
+
+	_vertex.SetXYZ(projectedVertexX, projectedVertexY, z);
+	const double norm = sqrt(dx*dx + dy*dy + 1);
+	_beam.SetXYZM(truthMomentum * dx / norm,
+	              truthMomentum * dy / norm,
+	              truthMomentum * 1 / norm,
```

Comment:
I'm not seeing how the "* 1" helps here.

Reply:
Just to make the three lines symmetric.
Should I remove it?